File: Python/Two Pointers/M-1004. Max Consecutive Ones III.py
class Solution:
    @classmethod
    def longestOnes(self, nums: list[int], k: int) -> int:
        cur_length, max_length = 0, 0
        l, r = 0, 0

        while r < len(nums):
            if nums[r] or k > 0:
                cur_length += 1
                max_length = max(max_length, cur_length)
                if not nums[r]:
                    k -= 1
                r += 1
                continue

            while k <= 0:
                cur_length -= 1
                if not nums[l]:
                    k += 1
                l += 1

        return max_length


testData = [
    {
        "nums": [1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0],
        "k": 2,
        "ans": 6,
    },
    {
        "nums": [0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1],
        "k": 3,
        "ans": 10,
    },
]

for data in testData:
    print(data["ans"] == Solution.longestOnes(data["nums"], data["k"]))

File: Two Pointers/M-1004. Max Consecutive Ones III/M-1004. Max Consecutive Ones III.ts
function longestOnes(nums: number[], k: number): number {
  let maxLength = 0;
  let curLength = 0;
  let [l, r] = [0, 0];

  while (r < nums.length) {
    if (nums[r] || k > 0) {
      curLength++;
      maxLength = Math.max(maxLength, curLength);
      if (!nums[r]) k--;

      r++;
      continue;
    }

    /** find the first zero from left */
    while (k <= 0) {
      curLength--;
      if (!nums[l]) k++;
      l++;
    }
  }

  return maxLength;
}

const testData = [
  {
    nums: [1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 0],
    k: 2,
    ans: 6,
  },
  {
    nums: [0, 0, 1, 1, 0, 0, 1, 1, 1, 0, 1, 1, 0, 0, 0, 1, 1, 1, 1],
    k: 3,
    ans: 10,
  },
];

for (const data of testData) {
  console.log(data['ans'] === longestOnes(data['nums'], data['k']));
}
